src/koopman.jl

```diff
@@ -16,25 +16,111 @@ end
 
 function expectation(g::Function, prob::ODEProblem, u0, p, expalg::Koopman, args...;
                         u0_CoV=(u,p)->u, p_CoV=(u,p)->p,
-                        maxiters=0,
+                        maxiters=1000000,
                         batch=0,
                         quadalg=HCubatureJL(),
                         ireltol=1e-2, iabstol=1e-2,
                         nout=1,kwargs...)
 
-    S = __make_map(prob, args...; kwargs...)
+    # S = __make_map(prob, args...; kwargs...)
 
-    expectation(g, S, u0, p, expalg, args...; u0_CoV=u0_CoV, p_CoV=p_CoV,
-                maxiters=maxiters, batch=batch,
-                ireltol=ireltol, iabstol=iabstol,
-                quadalg=quadalg,
-                nout=nout,kwargs...)
+    # expectation(g, S, u0, p, expalg, args...; u0_CoV=u0_CoV, p_CoV=p_CoV,
+    #             maxiters=maxiters, batch=batch,
+    #             ireltol=ireltol, iabstol=iabstol,
+    #             quadalg=quadalg,
+    #             nout=nout,kwargs...)
+
+    # construct extended state space
+    n_states = length(u0)
+    n_params = length(p)
+    ext_state = [u0; p]
+    # ext_state = (u0...,p...)
+
+    # find indices corresponding to distributions, check if sampleable and has non-zero support.
+    dist_mask = collect(isa.(ext_state, Sampleable) .& (minimum.(ext_state) .!= maximum.(ext_state)))
+    val_mask = .!(dist_mask)
+
+    # get distributions and indx in extended state space
+    dists = ext_state[dist_mask]
+
+    # create numerical state space values
+    # ext_state_val = minimum.(ext_state)
+    T0 = promote_type(eltype.(ext_state)...)
+    ext_state_val = [T0(minimum(v)) for v ∈ ext_state] |> collect  #collect needed for zygote for somereason. Otherwise is a tuple
```

Comment:
Ensures proper type promotion. This will also work for hyper-parameter optimization to distributions e.g. `truncated(Normal(ForwardDiff.Dual(0.0,1.0),3.0),-4,4)`